### PR TITLE
Modifications

### PR DESCRIPTION
**Pull request by ddavisgraphics**
1. is fixing an error in the add_project.env 
2. creating a command line tool for retrying errors

### CODE REVIEW

File: src/crons/retry-error.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+end
+
+# Grab the configuration files needed to rerun the file
+config=YAML.load_file("#{error_dir}/#{filename}")
+
+# check configs and enivronmental variables are the same
+if (config['project_name'] != ENV['HYDRA_PROJECT_NAME']) then
+  abort "Project name in control file does not match ENV HYDRA_PROJECT_NAME"
+end
+
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: You need to do the actual copy here. In the next steps you use a file in the `export` location but you don't ever put it back there.

File: src/crons/retry-error.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+Dir.chdir("/home/#{ENV['HYDRA_PROJECT_NAME']}.lib.wvu.edu/hydra/") do |dir_name|
+  export_locations = "/mnt/nfs-exports/mfcs-exports/#{config['project_name']}/export/#{config['time_stamp']}"
+  result = `/usr/local/bin/rails runner import/import.rb #{export_locations}`
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: You should carefully consider that this is actually what you want to do. 

I would argue that this file should be a "reset" command and not actually run the import. Its possible that some/all output could be lost, that you would otherwise see if you ran it by hand.

File: src/crons/retry-error.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  # looks to see if the result has exited with a false status
+  if (!$?.success?) then
+    # move from processing to error
+    # send email of failure and above the chron
+    send_notifications(config['contact_emails'], "Import of #{config['project_name']} failed.")
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: You don't want to send success emails out when you are troubleshooting an error.

File: src/crons/retry-error.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    abort "Error processing. Moved to error control directory. {#{result}}"
+  else
+    # otherwise import succeeded
+    send_notifications(config['contact_emails'], "Import of #{config['project_name']} succeeded.")
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: You really don't want to send out failures while troubleshooting. you may need to run this script many times before a success and this will alert the customers each time there is an error during testing.

File: src/crons/retry-error.rb [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    FileUtils.mv("#{error_dir}/#{config['time_stamp']}.yaml","#{success_dir}/#{config['time_stamp']}.yaml")
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: If you don't automatically run it, be sure you do this step by hand once you get a successful run.

File: src/crons/retry-error.rb [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+end
+
+# Don't let them add more arguments than needed
+if ARGV.length >= 2
+  puts "Error: Too many arguments supplied"
+  abort "Usage: ruby ./retry-error.rb {filename.yaml}"
+end
+
+# define the filename from the arguments
+filename = ARGV
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: this should be ARGV[0]

ARGV is an array. Ruby will autobox the first element to a string for you when you use it as a scalar like this. But this should really specify an index. That way if you ever need to add additional arguments later you don't have to go back and specify it.

File: src/crons/retry-error.rb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+# Mikes send notifications code
+def send_notifications(emailAddr, msgstr)
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: This method can be removed.

File: src/crons/retry-error.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  msgstr = "From: libsys@mail.wvu.edu\nSubject: Importing Update\n #{msgstr}"
+  Net::SMTP.start('smtp.wvu.edu', 25) do |smtp|
+    smtp.send_message msgstr, 'libsys@mail.wvu.edu', emailAddr
+  end
+end
+
+# Make sure directories are present and in vars for use later
+in_process_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/in-progress"
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: This variable is never used

File: src/crons/retry-error.rb [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+# Mikes send notifications code
+def send_notifications(emailAddr, msgstr)
+  msgstr = "From: libsys@mail.wvu.edu\nSubject: Importing Update\n #{msgstr}"
+  Net::SMTP.start('smtp.wvu.edu', 25) do |smtp|
+    smtp.send_message msgstr, 'libsys@mail.wvu.edu', emailAddr
+  end
+end
+
+# Make sure directories are present and in vars for use later
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: This comment is incorrect. you aren't checking that the directories exist. (There is no need to check for them either)

File: src/crons/retry-error.rb [hunk 10]
```diff
@@ -0,0 +1,89 @@
+#!/bin/ruby
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: This file should be moved to the utilities directory. It should never be run as a cron

File: src/crons/retry-error.rb [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+require 'yaml'
+require 'fileutils'
+require 'net/smtp'
```
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: not needed.

Comment (ddavisgraphics) [inline src/crons/retry-error.rb]: which part?  all of the requires appear to be used, I guess the email stuff could be chopped out since your doing it by hand on the command line anyway.

File: src/crons/retry-error.rb [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+#  retry-error
+# ==================================================
+# Name : David J. Davis
+# Date :  4/11/2017
+#
+# Requirements:  You must run add_project_env.sh first
+# so that the environmental variable is able to be used.
+#
+# Useage: From the terminal as root user run the command
+# ruby ./retry-error.rb {filename.yaml}
```
Comment (ddavisgraphics) [inline src/crons/retry-error.rb]: I debated on that for a while.  Should the extension be required in the commandline argument or should it just be hard coded?  I was thinking that the extension had to be in the argument so that if it did change the code wouldn't have to be changed, but looking at the code base it depends on yaml so either way the code base will have to be changed.

File: src/crons/retry-error.rb (hunk 6 above)
Comment (ddavisgraphics) [inline src/crons/retry-error.rb]: I know its an array and ruby auto maps arrays to number of variables.  `first, second, third = ARGV`
would all map correctly if used, empty ones would be empty or nil items.    

I feel this one is a matter of preference, but I'll change it.

File: src/crons/retry-error.rb [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  end
+end
+
+# Make sure directories are present and in vars for use later
+in_process_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/in-progress"
+error_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/error"
+success_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/finished"
+
+# if there is no file, abort the program
+if !File.exists?("#{error_dir}/#{filename}")
```
Comment (ddavisgraphics) [inline src/crons/retry-error.rb]: no its there.  The if doesn't need (), but the file exists method does because of the string being passed.

File: src/utils/retry-error.rb [hunk 14]
```diff
@@ -25,65 +24,44 @@
...
-# Make sure directories are present and in vars for use later
-in_process_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/in-progress"
+# define directories
 error_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/error"
 success_dir = "/mnt/nfs-exports/mfcs-exports/#{ENV['HYDRA_PROJECT_NAME']}/control/hydra/finished"
 
 # if there is no file, abort the program
-if !File.exists?("#{error_dir}/#{filename}")
-  abort "File is missing make sure that it is correct and present in the error folder"
+if !File.exist?("#{error_dir}/#{filename}.yaml")
```
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: `.yaml` needs to be removed, otherwise it will check if `12345.yaml.yaml` exists

File: src/utils/retry-error.rb [hunk 15]
```diff
@@ -25,65 +24,44 @@
...
+  abort '#{filename} is missing make sure that it is correct and present in the error folder'
 end
 
 # Grab the configuration files needed to rerun the file
-config=YAML.load_file("#{error_dir}/#{filename}")
+config = YAML.load_file("#{error_dir}/#{filename}.yaml")
```
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: remove `.yaml`, see above comment

File: src/utils/retry-error.rb [hunk 16]
```diff
@@ -25,65 +24,44 @@
 #  to the completed folder, if fails it is not moved.
 
 # Make Sure that the ENV variables are set
-if (!ENV.include? 'HYDRA_PROJECT_NAME') then
+if !ENV.include? 'HYDRA_PROJECT_NAME'
```
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: I hate rubo cop's styling with if/else blocks. IMO multi-line if should have parens and a then. But, if you are using rubo cop for linting, whatever. But. yuck. (imo).

Also, since you are following Rubocop style guide here, this should be a single line if (rubo cop prefers single line if whenever possible, which i do agree with :) )

`abort "Missing ENV variable, 'HYDRA_PROJECT_NAME'"` if !ENV.include? 'HYDRA_PROJECT_NAME'

File: src/crons/retry-error.rb (hunk 6 above)
Comment (MichaelRBond) [inline src/crons/retry-error.rb]: you didn't change it. :)

Perl does the same thing. 
`my($first, $second, $third) = @my_array;`

In some situations it makes sense. In some it doesn't.  I would argue that in this case it doesnt make sense. If your style is going to include the verbose commenting, it makes more sense to break it up. It makes your commenting easier and more intuitive. 

Also, it really isn't needed. You could just be using ARGV[0] instead of assigning the file name, since you already have the filename in a variable. 

essentially what you are dong is:
```
a = 1
b = a
puts b
```

File: src/utils/retry-error.rb [hunk 17]
```diff
@@ -25,65 +24,44 @@
...
 
 # Grab the configuration files needed to rerun the file
-config=YAML.load_file("#{error_dir}/#{filename}")
+config = YAML.load_file("#{error_dir}/#{filename}.yaml")
 
 # check configs and enivronmental variables are the same
-if (config['project_name'] != ENV['HYDRA_PROJECT_NAME']) then
-  abort "Project name in control file does not match ENV HYDRA_PROJECT_NAME"
+if config['project_name'] != ENV['HYDRA_PROJECT_NAME']
+  abort 'Project name in control file does not match ENV HYDRA_PROJECT_NAME'
```
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: why did you change these quotes to apostrophes? you are using quotes everywhere else, and there is nothing in this string that would need to be escaped in quotes.

File: src/utils/retry-error.rb (hunk 14 above)
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: since you appear to be adopting rubo cop this should be an unless instead of a negative if. also, one line.

`abort '#{filename} is missing make sure that it is correct and present in the error folder' unless File.exist?("#{error_dir}/#{filename}.yaml")`

This is another ruby style guide i don't agree with. I prefer the negative if ... but, one lining it with an if is good :)

Comment (ddavisgraphics) [inline src/utils/retry-error.rb]: All negatives should be an unless in that instance

Comment (ddavisgraphics) [inline src/utils/retry-error.rb]: Also it appears rubocop cusses at your line, says to switch to block syntax.

File: src/utils/retry-error.rb (hunk 17 above)
Comment (ddavisgraphics) [inline src/utils/retry-error.rb]: Because no variables are being parsed it prefers single quotes for strings where nothing is escaped or being parsed.

**Comment by ddavisgraphics** [comment]
Should be ready to go... let me know what your tests show.  I don't have anything setup locally to test it.

**Comment by MichaelRBond** [comment]
It looks good at a glance. I'll test it later today. If there are any issues from testing i'll correct them instead of doing any more back and forth dancing since you don't have a local env setup.

File: src/utils/retry-error.rb [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+unless ENV.include? 'HYDRA_PROJECT_NAME'
+  abort "Missing ENV variable, 'HYDRA_PROJECT_NAME'"
+end
+
+usage = 'Usage: ruby ./retry-error.rb {filename}'
+abort "Error: Expected filename.\n #{usage}" if ARGV.empty?
+abort "Error: Too many arguments.\n #{usage}" if ARGV.length > 1
+
+filename = ARGV
```
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: So, this doesn't work. It comes back with: 
```
ruby retry-error.rb 1490476034.yaml
["1490476034.yaml"] does not exist
```
Setting it to argv[0] does work

File: src/utils/retry-error.rb [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+end
+
+config = YAML.load_file("#{error_dir}/#{filename}")
+
+if config['project_name'] != ENV['HYDRA_PROJECT_NAME']
+  abort 'Project name in control file does not match ENV HYDRA_PROJECT_NAME'
+end
+
+Dir.chdir("/home/#{ENV['HYDRA_PROJECT_NAME']}.lib.wvu.edu/hydra/") do |dir_name|
+  export_locations = "/mnt/nfs-exports/mfcs-exports/#{config['project_name']}/export/#{config['time_stamp']}"
```
Comment (MichaelRBond) [inline src/utils/retry-error.rb]: This isn't the correct location. You never move it to the export directory, and you reference it from the error directory below.

**Comment by MichaelRBond** [comment]
This pull request is ready to be merged in. @ddavisgraphics go ahead and click the button when you are ready to work with it. 

I can the script locally, and the import works as expected (I do not have the newest changes to hydra-wvcp pulled down).
